core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -462,6 +484,22 @@ protected Builder buildAggregate(Aggregate e, Builder builder,
       // as there is at least one aggregate function.
       builder.setGroupBy(new SqlNodeList(groupByList, POS));
     }
+
+    if (builder.clauses.contains(Clause.HAVING) && !e.getGroupSet()
+        .equals(ImmutableBitSet.union(e.getGroupSets()))) {
+      // groupSet contains at least one column that is not in any groupSets.
+      // To make such columns must appear in the output (their value will
+      // always be NULL), we generate an extra grouping set, then filter
+      // it out using a "HAVING GROUPING_ID(groupSets) <> 0".
+      // We want to generate the
+      final SqlNodeList groupingList = new SqlNodeList(POS);
+      e.getGroupSet().forEach(g ->
+          groupingList.add(builder.context.field(g)));
+      SqlNumericLiteral zero = SqlNumericLiteral.createExactNumeric(String.valueOf(0), POS);
+      SqlNode condition = SqlStdOperatorTable.NOT_EQUALS.createCall(POS,
+          SqlStdOperatorTable.GROUPING_ID.createCall(groupingList), zero);
```

Comment:
nit: This should be `GROUPING` which is now equivalent to `GROUPING_ID` except the former is standardized in ANSI SQL. See [CALCITE-1652](https://issues.apache.org/jira/browse/CALCITE-1652)

Reply:
yes, `GROUPING` is preferred, even though they are equivalent.

Reply:
done